Typescript

craigmulligan opened the pull request:
~~Still have to get the tests running but ofc jest is being a pain. Feel free to jump in if you like @zachblume, otherwise I'll get to it later this week.~~

Okay, I replaced jest with vitest it's *just works*. 

Fixes #59

zachblume commented:
Looks great so far and I think we can merge it in. My typescript knowlege is largely confined to application development; I just picked up my knowlege of JS necessary for getting packages to NPM for this project. I was surprised to find it so simple. 

1. Does this follow the general conventions of typescript use for npm packages?
2. Are we exporting types for this package correctly with this implementation? When I brought in `autospec` to a web ui repo, I had to define a .d.ts to use it.
3. `build && npx package` and using esbuild instead of tsc for building the js caught my eye -- just want to double check with you
4. In my prior typescript experience, I've turned on typescript strict mode and added @typescript-eslint/no-explicit-any to prevent future escape

craigmulligan commented:
1. With most things in the JS ecosystem there isn't a lot in the way of conventions. Especially with regard to the output format. I've kept things as is in that the output is a esm format (not commonjs). If someone needs commonjs we'd need to add a tsconfig for that specifically, but as far as I can tell everyone is trying to move away from commonjs to esm.
2. Good catch, I think we just needed to define them in the package.json and make sure tsc is outputting the .dts files. Should be working now?
3. Yeah, I've seen many people recommend esbuild for it's simplicity and speed, but it does add more tooling that we don't strictly need. I've switched it back to using tsc for both typechecking and transpiling. 
4. Yeah, I was hoping to turn on strict mode, but there were just too many errors and ran out of time, I think it's something we can work towards rather than knock out right now. But good call on eslint, I forgot about that. I've updated it to handle typescript, it has the `@typescript-eslint/no-explicit-any` built in.
